### PR TITLE
[UNIT-TEST] Validator Exchange Rate Callback

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## Brief Changelog
- Added unit test for validator exchange rate callback
- Minor variable renaming in callback function
- Fixed bug in submit tx (the wrong epoch type was used for the TTL)

## Author's Checklist

I have...

- [x] Run and PASSED locally all GAIA integration tests
- [ ] If the change is contentful, I either:
    - [X] Added a new unit test OR 
    - [ ] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [X] not documented  (tests)

### CODE REVIEW

File: x/stakeibc/keeper/callbacks.go [hunk 1]
```diff
@@ -194,69 +194,89 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
...
+		return sdkerrors.Wrapf(types.ErrMarshalFailure, errMsg)
 	}
-	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v queriedValidator %v", zone.ChainId, queriedValidator))
+	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v queriedValidator %v", hostZone.ChainId, queriedValidator))
 
 	// ensure ICQ can be issued now! else fail the callback
-	valid, err := k.IsWithinBufferWindow(ctx)
+	withinBufferWindow, err := k.IsWithinBufferWindow(ctx)
 	if err != nil {
+		// QUESTION: this is the case where the user submitted the query within the buffer window,
```
Comment (asalzmann) [inline x/stakeibc/keeper/callbacks.go:225]: It looks like the errors that could surface here are (1) int casts gone wrong (2) epoch trackers not being found

**Review by asalzmann — APPROVED** [review]
lgtm! would be good if @riley-stride took a look at this one as well

File: x/stakeibc/keeper/callbacks.go [hunk 2]
```diff
@@ -194,69 +194,89 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
...
+		// but by the time it got back, we were outside of the window
+		// Is there anything we should do differently here. Not a great UX for the TX to look successful but fail under the hood
```
Comment (riley-stride) [inline x/stakeibc/keeper/callbacks.go:227]: Great question. Unfortunately not sure we can do anything differently, since errors surfaced to the user in response to their tx must be knowable atomically, whereas this can't be known until after one daisy chain

File: x/stakeibc/keeper/callbacks.go [hunk 3]
```diff
@@ -194,69 +194,89 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
...
 	}
 	// get the stride epoch number
 	strideEpochTracker, found := k.GetEpochTracker(ctx, epochtypes.STRIDE_EPOCH)
 	if !found {
 		k.Logger(ctx).Error("failed to find stride epoch")
 		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no epoch number for epoch (%s)", epochtypes.STRIDE_EPOCH)
 	}
-	// converting 1.0 gives us the exchange rate to later use in the next CB
-	v.InternalExchangeRate = &types.ValidatorExchangeRate{
+	// We want the validator's internal exchange rate which is held internally behind `validator.TokensFromShares`
```
Comment (riley-stride) [inline x/stakeibc/keeper/callbacks.go:246]: Thank you for documenting all this and clearing up variable names, looks MUCH better 🙏

File: x/stakeibc/keeper/msg_server_submit_tx.go [hunk 4]
```diff
@@ -373,7 +373,7 @@ func (k Keeper) QueryValidatorExchangeRate(ctx sdk.Context, msg *types.MsgUpdate
 	data := stakingtypes.GetValidatorKey(valAddr)
 
 	// get ttl
-	ttl, err := k.GetStartTimeNextEpoch(ctx, epochstypes.DAY_EPOCH)
+	ttl, err := k.GetStartTimeNextEpoch(ctx, epochstypes.STRIDE_EPOCH)
```
[the remark this answers is not in the document]
Comment (riley-stride) [reply x/stakeibc/keeper/msg_server_submit_tx.go:376]: +100! great catch @sampocs

File: x/stakeibc/keeper/icqcallbacks_validator_test.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+type ValidatorICQCallbackArgs struct {
+	query        icqtypes.Query
+	callbackArgs []byte
+}
+
+type ValidatorICQCallbackTestCase struct {
+	initialState         ValidatorICQCallbackState
+	validArgs            ValidatorICQCallbackArgs
+	valAddress           string
```
Comment (riley-stride) [inline x/stakeibc/keeper/icqcallbacks_validator_test.go:28]: nit: this is unused, either remove or use tx.ValAddress in ln 47

File: x/stakeibc/keeper/callbacks.go [hunk 6]
```diff
@@ -194,69 +194,89 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
...
+	//  Since,
+	//     exchange_rate = num_tokens / num_shares
+	//  We can use `validator.TokensFromShares`, plug in 1.0 for the number of shares,
+	//    and the returned number of tokens will be equal to the internal exchange rate
+	validator.InternalExchangeRate = &types.ValidatorExchangeRate{
 		InternalTokensToSharesRate: queriedValidator.TokensFromShares(sdk.NewDec(1.0)),
```
Comment (riley-stride) [inline x/stakeibc/keeper/callbacks.go:252]: Could we add a test for the case that tokens and shares returned are both 0? Do we get a divide by 0 here?

**Review by riley-stride — APPROVED** [review]
Awesome test and cleanup of this callback. Only suggestion is we add a test for the case where a validator with 0 stake is queried, I worry we might get a division by zero, might need to catch that case.

File: x/stakeibc/keeper/callbacks.go (hunk 1 above)
Comment (sampocs) [reply x/stakeibc/keeper/callbacks.go:225]: ah good eye! You're right I read it wrong

File: x/stakeibc/keeper/icqcallbacks_validator_test.go (hunk 5 above)
Comment (sampocs) [reply x/stakeibc/keeper/icqcallbacks_validator_test.go:28]: removed!

File: x/stakeibc/keeper/callbacks.go (hunk 6 above)
Comment (sampocs) [reply x/stakeibc/keeper/callbacks.go:252]: Great call! I think the case you're looking for is when DelegatorShares is 0. Cause shares is hard coded to 1.0 and tokens is in the numerator. Does that sound right?

Comment (sampocs) [reply x/stakeibc/keeper/callbacks.go:252]: Just tested it and it would have panic'd!!! Great catch!!

File: x/stakeibc/keeper/callbacks.go [hunk 7]
```diff
@@ -243,6 +243,15 @@ func ValidatorExchangeRateCallback(k Keeper, ctx sdk.Context, args []byte, query
 		k.Logger(ctx).Error("failed to find stride epoch")
 		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no epoch number for epoch (%s)", epochtypes.STRIDE_EPOCH)
 	}
+
+	// If the validator's delegation amount is 0, we'll get a division by zero error when trying to get the exchange rate
+	// Because
```
Comment (riley-stride) [inline x/stakeibc/keeper/callbacks.go:248]: nit: finish comment

**Review by riley-stride — APPROVED** [review]
looks great! nice tests

File: x/stakeibc/keeper/callbacks.go (hunk 7 above)
Comment (sampocs) [reply x/stakeibc/keeper/callbacks.go:248]: done
